applications/HDF5Application/custom_io/hdf5_partitioned_model_part_io.cpp

```diff
@@ -68,103 +90,109 @@ void PartitionedModelPartIO::WriteNodes(NodesContainerType const& rNodes)
+    block_for_each(r_nodes, [my_pid](auto& rNode) {
+        rNode.FastGetSolutionStepValue(PARTITION_INDEX) = my_pid;
+    });
 
-    // Read and assign partition ids for ghosts. If this is slow, try storing the
-    // partition index in the node's non-historical container when the ghost nodes and
-    // local nodes are read. Then copy it to the solution step data after the buffer
-    // is initialized.
+    // now we read the ghost node partition index from the H5 file.
+    const auto& r_ghost_node_path = mPrefix + "/Nodes/Ghost";
     unsigned start_index, block_size;
-    std::tie(start_index, block_size) = StartIndexAndBlockSize(rPath);
+    std::tie(start_index, block_size) = StartIndexAndBlockSize(*mpFile, r_ghost_node_path);
     Vector<int> partition_ids, node_ids;
-    mpFile->ReadDataSet(rPath + "/PARTITION_INDEX", partition_ids, start_index, block_size);
-    mpFile->ReadDataSet(rPath + "/Ids", node_ids, start_index, block_size);
-    for (unsigned i = 0; i < node_ids.size(); ++i)
-        r_nodes[node_ids[i]].FastGetSolutionStepValue(PARTITION_INDEX) = partition_ids[i];
+    mpFile->ReadDataSet(r_ghost_node_path + "/PARTITION_INDEX", partition_ids, start_index, block_size);
+    mpFile->ReadDataSet(r_ghost_node_path + "/Ids", node_ids, start_index, block_size);
+
+    // now we correct PARTITION_INDEX of ghost nodes.
+    IndexPartition<IndexType>(node_ids.size()).for_each([&r_nodes, &node_ids, &partition_ids](const auto Index) {
+        r_nodes[node_ids[Index]].FastGetSolutionStepValue(PARTITION_INDEX) = partition_ids[Index];
+    });
 
     KRATOS_CATCH("");
 }
 
-std::tuple<unsigned, unsigned> PartitionedModelPartIO::StartIndexAndBlockSize(std::string const& rPath) const
+void PartitionedModelPartIO::SetCommunicator(ModelPart& rModelPart) const
 {
-    KRATOS_TRY;
-    unsigned start_index, block_size;
-    std::tie(start_index, block_size) = HDF5::StartIndexAndBlockSize(*mpFile, rPath);
-    return std::make_tuple(start_index, block_size);
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(mpFile->GetDataCommunicator().IsDistributed())
+        << "Trying to use PartitionedModelPartIO with an HDF5File which is based on serial communicator.";
+
+    // set the mpi communicator
+    if (!rModelPart.GetCommunicator().IsDistributed()) {
+        rModelPart.SetCommunicator(Kratos::make_shared<MPICommunicator>(rModelPart.pGetNodalSolutionStepVariablesList().get(), mpFile->GetDataCommunicator()));
+    }
+
+    // get the communicator
+    auto& r_communicator = rModelPart.GetCommunicator();
+    const auto my_pid = r_communicator.MyPID();
+
+    // now we have set all the PARTITION_INDEX values correctly for all nodes,
+    // now we set the local mesh and ghost mesh nodes correctly.
+    auto& r_local_nodes = r_communicator.LocalMesh().Nodes();
+    auto& r_ghost_nodes = r_communicator.GhostMesh().Nodes();
+
+    for (auto& p_node : rModelPart.Nodes().GetContainer()) {
```

Comment:
yikes

I hope `PointerVectorSet::GetContainer` is going away, so it'd be best to think ahead and use iterators instead.

```suggestion
    for (auto it_node=rModelPart.Nodes().ptr_begin(); it_node!=rModelPart.Nodes().ptr_end(); ++it_node) {
        const auto& rp_node = *it_node;
```